core/drop_test.go

```diff
@@ -1,19 +1,222 @@
 package core
 
 import (
+	"bytes"
+	"github.com/0xPolygon/go-ibft/messages"
+	"github.com/0xPolygon/go-ibft/messages/proto"
+	"math/rand"
 	"testing"
 	"time"
 
 	"github.com/stretchr/testify/assert"
 )
 
+func TestDropAllAndRecover(t *testing.T) {
+	t.Parallel()
+
+	var (
+		numNodes       = uint64(6)
+		insertedBlocks = make([][]byte, numNodes)
+	)
+
+	cluster := newCluster(
+		numNodes,
+		func(c *cluster) {
+			for nodeIndex, node := range c.nodes {
+				i := nodeIndex
+				currentNode := node
+				node.core = NewIBFT(
+					mockLogger{},
+					&mockBackend{
+						isValidBlockFn:         isValidProposal,
+						isValidProposalHashFn:  isValidProposalHash,
+						isValidSenderFn:        nil,
+						isValidCommittedSealFn: nil,
+						isProposerFn:           c.isProposer,
+
+						idFn: node.addr,
+
+						buildProposalFn:           buildValidProposal,
+						buildPrePrepareMessageFn:  node.buildPrePrepare,
+						buildPrepareMessageFn:     node.buildPrepare,
+						buildCommitMessageFn:      node.buildCommit,
+						buildRoundChangeMessageFn: node.buildRoundChange,
+
+						insertBlockFn: func(proposal []byte, _ []*messages.CommittedSeal) {
+							insertedBlocks[i] = proposal
+						},
+						hasQuorumFn: c.hasQuorumFn,
+					},
+					&mockTransport{multicastFn: func(message *proto.Message) {
+						if currentNode.offline {
+							return
+						}
+
+						c.gossip(message)
+					}},
+				)
+			}
+		},
+	)
+
+	// Progress the chain to claim it works ok by default
+	err := cluster.progressToHeight(5*time.Second, 1)
+	assert.NoError(t, err, "unable to reach height: %w", err)
+	assert.Equal(t, uint64(1), cluster.latestHeight)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+
+	insertedBlocks = make([][]byte, numNodes) // Purge
+
+	// Stop all nodes and make sure no blocks are written
+	cluster.stopN(len(cluster.nodes))
+	cluster.progressToHeight(2*time.Second, 2)
+	assertNoInsertedBlocks(insertedBlocks, t)
+
+	// Start all and expect valid blocks to be written again
+	cluster.startN(len(cluster.nodes))
+	cluster.progressToHeight(5*time.Second, 10)
+	assertValidInsertedBlocks(insertedBlocks, t) // Make sure the inserted blocks are valid
+}
+
+func assertNoInsertedBlocks(blocks [][]byte, t *testing.T) {
+	t.Helper()
+
+	assertNInsertedBlocks(0, blocks, t)
+}
+
+func assertNInsertedBlocks(n int, blocks [][]byte, t *testing.T) {
```

Comment:
What calls `assertNInsertedBlocks` except for `assertNoInsertedBlocks`? You can check in ` assertNoInsertedBlocks` directly.

Reply:
There are some cases